hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpTokenAssociationsSubcommand.java

```diff
@@ -0,0 +1,267 @@
+
+        writer.write(
+                "# === %8d token associations (%d the hard way), %7d accounts with associations, %7d token types associated with accounts%n"
+                        .formatted(
+                                tokenAssociations.size(),
+                                totalTokenRelsByUniques,
+                                uniqueAccounts.size(),
+                                uniqueTokens.size()));
+
+        final var tokensByAccountsHistogram = tokensByAccounts.values().stream()
+                .collect(Collectors.groupingBy(n -> 0 == n ? 0 : (int) Math.log10(n), Collectors.counting()));
+        final int maxDigits = tokensByAccountsHistogram.keySet().stream()
+                .max(Comparator.naturalOrder())
+                .orElse(0);
+
+        final var sb = new StringBuilder(1000);
+        sb.append("#   === token associations per account histogram ===%n".formatted());
+        for (int i = 0; i <= maxDigits; i++) {
+            sb.append("# %11s: %8d%n"
+                    .formatted("<=" + (int) Math.pow(10, i), tokensByAccountsHistogram.getOrDefault(i, 0L)));
+        }
+        writer.write(sb);
+        writer.write("");
+    }
+
+    @NonNull
+    String formatHeader() {
+        return "account;token;balance;frozen;kycGranted;automaticAssociation;prev;next";
+    }
+
+    @NonNull
+    String toCsv(final boolean b) {
+        return b ? "T" : "";
+    }
+
+    @NonNull
+    String formatTokenRel(@NonNull final TokenRel t) {
+        return "%d;%d;%d;%s;%s;%s;%d;%d"
+                .formatted(
+                        t.account(),
+                        t.tokenId(),
+                        t.balance(),
+                        toCsv(t.isFrozen()),
+                        toCsv(t.isKycGranted()),
+                        toCsv(t.isAutomaticAssociation()),
+                        t.prev(),
+                        t.next());
+    }
+
+    void reportOnTokenAssociations(
+            @NonNull final Writer writer, @NonNull final SortedMap<Pair<Long, Long>, TokenRel> tokenAssociations) {
+        requireNonNull(writer, "writer");
+        requireNonNull(tokenAssociations, "tokenAssociations");
+        writer.writeln(formatHeader());
+        tokenAssociations.forEach((ignored, value) -> writer.writeln(formatTokenRel(value)));
+        writer.writeln("");
+    }
+
+    @NonNull
+    SortedMap<Pair<Long, Long>, TokenRel> gatherTokenAssociations(@NonNull TokenRelStorageAdapter tokenRelStore) {
+        requireNonNull(tokenRelStore, "tokenRelStore");
+
+        final SortedMap<Pair<Long, Long>, TokenRel> r = new TreeMap<>(getPairOfLongsComparator());
+
+        if (tokenRelStore.areOnDisk()) {
+            final var tokenRels = requireNonNull(tokenRelStore.getOnDiskRels());
+            final var tokenAssociations = new ConcurrentLinkedQueue<TokenRel>();
+            try {
+                final var threadCount = 8; // Good enough for my laptop, why not?
```

Comment:
May be good to make this configurable at some point

Reply:
+1

Reply:
I'll do it in several of these dumpers at the same time, separate PR.